### ISSUE TITLE
nIndex maximum interfaces

### ISSUE THREAD

**Issue by mmatuska79** [ntop/ntopng#2558]
Hello, 
   we have Enterprise license for ntopng. I have configured nIndex as data flow dump, but I get following messages during startup:

`ERROR: The maximum number of interfaces that can be used with nIndex is 8.`

We have approx 15-20 interfaces. Is there a possibility to increase the limit?
I have also an option to use ElasticSearch ad different data dump. What would you recommend (nIndex or Elastic)?

Thanks, Mirek

`v.3.9.190506    [Enterprise/Professional build]`
`GIT rev:        dev:5bc188b1985291eab13a03c4473d54a309afec87:20190506`
`Pro rev:        r2321`
`Built on:`
`System Id:      68B9F75F76066ACF`
`Platform:       x86_64`
`Edition:        Enterprise`
`License Type:   Permanent`
`Maintenance:    Until Sat May  2 14:35:20 2020 [360 days left]`

**Comment by emanuele-f** [comment]
Hi, Elastic is only supported for exporting flows, ntopng will not be able to read them back and show into the gui. Regarding the limit, we will discuss the implications of increasing it.

**Comment by mmatuska79** [comment]
Thanks for the answer. 
So if you manage to increase the limit for nIndex interfaces (24 would be OK for one of our implementation scenario), we will be happy using them rather than Elastic.
Mirek

**Comment by gianogli** [comment]
We noticed the same problem. We are using 9 interfaces now but we have only 7 `nIndex` timeseries (not 8) due to these limitation:
`ERROR: The maximum number of interfaces that can be used with nIndex is 8.`
_

```
# ntopng --version
v.3.8.190307    [Enterprise/Professional build]
GIT rev:        3.8-stable:b2878274dc6c6b554f642ebe8bc90a362ed43585:20190307
Pro rev:        r2171
```

**Comment by emanuele-f** [comment]
It is currently possible to disable flows export for specific interfaces from the interface settings https://www.ntop.org/guides/ntopng/web_gui/interfaces.html#settings . There are implications in raising the maximum interfaces limit right now, I'll mark this as an enhancement.

**Comment by simonemainardi** [comment]
Currently, we cannot extend the limit of 8 nIndex interfaces per ntopng instance. This is due to the number of files open which would be too high for a single process.

However, one can start multiple ntopng instances, each one monitoring at most 8 interfaces. To start multiple instances on the same machine, one can just run multiple `ntopng`s and specify different data directories (option -d) and different redis databases (option -r).

@gianogli the fact that you were seeing 7 was a bug and has been fixed.

**Comment by gianogli** [comment]
> Currently, we cannot extend the limit of 8 nIndex interfaces per ntopng instance. This is due to the number of files open which would be too high for a single process.
> 
> However, one can start multiple ntopng instances, each one monitoring at most 8 interfaces. To start multiple instances on the same machine, one can just run multiple `ntopng`s and specify different data directories (option -d) and different redis databases (option -r).
> 
> @gianogli the fact that you were seeing 7 was a bug and has been fixed.

Hi Simone,
Are you sure that the "7" bug has been fixed?
Today I was finally able to remove one of my interface (now I've 8 interfaces) but when I start ntopng I have:
```
04/Oct/2019 11:16:16 [NetworkInterface.cpp:6696] ERROR: nIndex cannot be enabled for zc:2@1.
04/Oct/2019 11:16:16 [NetworkInterface.cpp:6698] ERROR: The maximum number of interfaces that can be used with nIndex is 8.
04/Oct/2019 11:16:16 [NetworkInterface.cpp:6701] ERROR: Interface will continue to work without nIndex support.
04/Oct/2019 11:16:16 [NetworkInterface.cpp:6696] ERROR: nIndex cannot be enabled for zc:2@2.
04/Oct/2019 11:16:16 [NetworkInterface.cpp:6698] ERROR: The maximum number of interfaces that can be used with nIndex is 8.
04/Oct/2019 11:16:16 [NetworkInterface.cpp:6701] ERROR: Interface will continue to work without nIndex support.
04/Oct/2019 11:16:16 [NetworkInterface.cpp:6696] ERROR: nIndex cannot be enabled for zc:2@3.
04/Oct/2019 11:16:16 [NetworkInterface.cpp:6698] ERROR: The maximum number of interfaces that can be used with nIndex is 8.
04/Oct/2019 11:16:16 [NetworkInterface.cpp:6701] ERROR: Interface will continue to work without nIndex support.
04/Oct/2019 11:16:16 [NetworkInterface.cpp:6696] ERROR: nIndex cannot be enabled for zc:2@4.
04/Oct/2019 11:16:16 [NetworkInterface.cpp:6698] ERROR: The maximum number of interfaces that can be used with nIndex is 8.
04/Oct/2019 11:16:16 [NetworkInterface.cpp:6701] ERROR: Interface will continue to work without nIndex support.
04/Oct/2019 11:16:16 [NetworkInterface.cpp:6696] ERROR: nIndex cannot be enabled for zc:2@5.
04/Oct/2019 11:16:16 [NetworkInterface.cpp:6698] ERROR: The maximum number of interfaces that can be used with nIndex is 8.
04/Oct/2019 11:16:16 [NetworkInterface.cpp:6701] ERROR: Interface will continue to work without nIndex support.
04/Oct/2019 11:16:16 [NetworkInterface.cpp:6696] ERROR: nIndex cannot be enabled for zc:2@6.
04/Oct/2019 11:16:16 [NetworkInterface.cpp:6698] ERROR: The maximum number of interfaces that can be used with nIndex is 8.
04/Oct/2019 11:16:16 [NetworkInterface.cpp:6701] ERROR: Interface will continue to work without nIndex support.
04/Oct/2019 11:16:16 [NetworkInterface.cpp:6696] ERROR: nIndex cannot be enabled for zc:2@7.
04/Oct/2019 11:16:16 [NetworkInterface.cpp:6698] ERROR: The maximum number of interfaces that can be used with nIndex is 8.
04/Oct/2019 11:16:16 [NetworkInterface.cpp:6701] ERROR: Interface will continue to work without nIndex support.
04/Oct/2019 11:16:16 [NetworkInterface.cpp:6696] ERROR: nIndex cannot be enabled for zc:2@8.
04/Oct/2019 11:16:16 [NetworkInterface.cpp:6698] ERROR: The maximum number of interfaces that can be used with nIndex is 8.
04/Oct/2019 11:16:16 [NetworkInterface.cpp:6701] ERROR: Interface will continue to work without nIndex support.
```

I think that the problem is here:
https://github.com/ntop/ntopng/blob/583eeaae3e6149bf75dca0c1baf67d875c0d4750/src/NetworkInterface.cpp#L6704

On "dev" branch the same line is:
https://github.com/ntop/ntopng/blob/bee1efddedbf20c6ae9f0a8a0b10c5c7017feccb/src/NetworkInterface.cpp#L7096

Could you verify it and eventually backport the fix to the "3.8-stable" branch?

Thanks... :-)
